### PR TITLE
Implement #evaluate in every condition

### PR DESCRIPTION
**Pull request by migl**
* Add documentation and make it consistent across classes

### CODE REVIEW

File: lib/ruy/conditions/assert.rb [hunk 1]
```diff
@@ -1,20 +1,30 @@
...
+    #
     class Assert < Condition
-      attr_reader :attr
+      attr_reader :key
 
-      # @param attr Context attribute's attr
-      def initialize(*attrs)
+      # @param key
+      # @example evaluate that :key passes
+      #   Asset.new(:key)
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/assert.rb]: Shoud be Assert instead of Asset

File: lib/ruy/conditions/assert.rb [hunk 2]
```diff
@@ -1,20 +1,30 @@
...
+      def initialize(*keys)
         super
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/assert.rb]: I believe if keys.any? shouldn't be necessary. @key will be nil unless keys.any? anyway.

File: lib/ruy/conditions/between.rb [hunk 3]
```diff
@@ -1,28 +1,27 @@
 module Ruy
   module Conditions
 
-    # Expects that a context value belongs to a given range.
+    # Expects that a context key stores a value between a defined range
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/between.rb]: Is it right to say: between a range? I would prefer: included in a range.

File: lib/ruy/conditions/between.rb [hunk 4]
```diff
@@ -1,28 +1,27 @@
...
-      attr_reader :from, :to, :range, :attr
+      attr_reader :from, :to, :range, :key
 
-      # @overload initialize(from, to, attr)
+      # @overload initialize(from, to, key)
       #   @param from Range lower bound
       #   @param to Range upper bound
-      #   @param attr Name of the attribute that will be evaluated
-      # @overload initialize(range, attr)
+      #   @param key Name of the attribute that will be evaluated
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/between.rb]: Shouldn't be attribute.
 
Alternative:
Name of the key for which its value will be evaluated

File: lib/ruy/conditions/assert.rb [hunk 5]
```diff
@@ -1,20 +1,30 @@
...
       end
 
       def ==(o)
         o.kind_of?(Assert) &&
-          o.attr == @attr
+        o.key == @key
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/assert.rb]: Should be indented.

File: lib/ruy/conditions/condition.rb [hunk 6]
```diff
@@ -9,30 +13,39 @@ def initialize(*params)
         @params = params
       end
 
-      # Evaluates a context value to check whether or not it satisfies the condition.
+      # Evaluates a context's object for its compliance with the condition.
       #
       # @param [Context] ctx
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/condition.rb]: Use the new syntax:

# @param request [Request] the request object

File: lib/ruy/conditions/compound_condition.rb [hunk 7]
```diff
@@ -8,19 +12,32 @@ def initialize(*params)
...
+
+      protected
+
+      # Evaluates a context to check if it satisfies the set of sub-conditions
       #
-      # @return [true] When all the conditions are satisfied
-      # @return [false] Otherwise
-      def call(ctx)
+      # @abstract
+      # @param [Context]
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/compound_condition.rb]: Param name missing. Keep the same style than the other files.

File: lib/ruy/conditions/compound_condition.rb [hunk 8]
```diff
@@ -8,19 +12,32 @@ def initialize(*params)
...
         end
       end
 
-      def ==(o)
-        o.kind_of?(self.class) &&
-          o.conditions == @conditions
+      # Lookups the object of evaluation in the context
+      # based on the initialization attribute
+      #
+      # @param [Context]
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/compound_condition.rb]: The same as above.

File: lib/ruy/conditions/condition.rb [hunk 9]
```diff
@@ -9,30 +13,39 @@ def initialize(*params)
...
+      #
+      # @abstract
+      # @param obj
+      # @return [Boolean] true if it satisfies the condition, false otherwise
+      def evaluate(obj); end
+
+      # Lookups the object of evaluation in the context
+      # based on the initialization attribute
       #
       # @param [Context] ctx
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/condition.rb]: Same as above.

File: lib/ruy/conditions/condition.rb [hunk 10]
```diff
@@ -9,30 +13,39 @@ def initialize(*params)
...
-      # @return [Object]
-      def solve(ctx)
-        defined?(@attr) ? ctx.resolve(@attr) : ctx.values
+      # @return an object stored in the context
+      def resolve(ctx)
+        has_key? ? ctx.resolve(key) : ctx.object
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/condition.rb]: ctx.resolve(@key) ?

File: lib/ruy/conditions/day_of_week.rb [hunk 11]
```diff
@@ -4,34 +4,40 @@
...
         super
         @day = day
-        @attr = attr
+        @key = key
         @tz_identifier = tz_identifier
       end
 
       def ==(o)
         o.kind_of?(DayOfWeek) &&
           o.day == @day &&
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/day_of_week.rb]: Why is it indented this way?

File: lib/ruy/conditions/day_of_week.rb [hunk 12]
```diff
@@ -4,34 +4,40 @@
...
-          o.attr == @attr &&
+          o.key == @key &&
           o.tz_identifier == @tz_identifier
       end
 
       protected
 
-      def evaluate(value)
+      def evaluate(date)
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/day_of_week.rb]: I would like documentation for this.

File: lib/ruy/conditions/dig.rb [hunk 13]
```diff
@@ -1,32 +1,40 @@
 module Ruy
   module Conditions
 
-    # Digs into a Hash allowing to define conditions over nested attributes
+    # Navigates into a Hash allowing to define conditions over nested attributes
+    #
     class Dig < CompoundCondition
       attr_reader :chain
 
       # @param *chain Context attributes' chain
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/dig.rb]: Context attribute's chain ?

File: lib/ruy/conditions/dig.rb [hunk 14]
```diff
@@ -1,32 +1,40 @@
...
         @chain = chain
       end
 
-      def call(ctx)
-        leafctx = @chain.reduce(ctx) do |currentctx, attr|
-          if currentctx.include?(attr)
-            Ruy::Context.new(currentctx.resolve(attr))
+      def ==(o)
+        super &&
+        o.chain == @chain
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/dig.rb]: Indentation.

File: lib/ruy/conditions/eq.rb [hunk 15]
```diff
@@ -1,22 +1,25 @@
...
-      def initialize(obj, *attrs)
+      # @param obj
+      # @param key
+      # @example evaluate that :key is 5
+      #   Eq.new(5, :key)
+      def initialize(obj, *keys)
         super
         @obj = obj
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/eq.rb]: The same as in the other example. If keys.any? shouldn't be necessary.

File: lib/ruy/conditions/eq.rb [hunk 16]
```diff
@@ -1,22 +1,25 @@
...
       end
 
       def ==(o)
         o.kind_of?(Eq) &&
           o.obj == @obj &&
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/eq.rb]: Indentation. Keep the same criteria.

File: lib/ruy/conditions/every.rb [hunk 17]
```diff
@@ -1,28 +1,35 @@
...
 
-      # @param attr Context attribute's name
-      def initialize(*attrs)
+      # @param key
+      # @example check that every element of :array matches the sub-conditions
+      #   Every.new(:array)
+      def initialize(*keys)
         super
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/every.rb]: The same as above.

File: lib/ruy/conditions/greater_than.rb [hunk 18]
```diff
@@ -1,22 +1,25 @@
...
-      # @param attr Context attribute's name
-      def initialize(obj, *attrs)
+      # @param key
+      # @example expects that :key is greater than 5
+      #   GreaterThan.new(5, :key)
+      def initialize(obj, *keys)
         super
         @obj = obj
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/greater_than.rb]: The same as in the other cases.

File: lib/ruy/conditions/greater_than.rb [hunk 19]
```diff
@@ -1,22 +1,25 @@
...
       end
 
       def ==(o)
         o.kind_of?(GreaterThan) &&
           o.obj == @obj &&
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/greater_than.rb]: The same as in the other cases.

File: lib/ruy/conditions/greater_than_or_equal.rb [hunk 20]
```diff
@@ -1,22 +1,26 @@
...
-      # @param attr Context attribute's name
-      def initialize(obj, *attrs)
+      # @param key
+      # @example check that :key is greater than or equal to 5
+      #   GreaterThanOrEqual.new(5, :key)
+      def initialize(obj, *keys)
         super
         @obj = obj
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/greater_than_or_equal.rb]: The same as in the other cases.

File: lib/ruy/conditions/greater_than_or_equal.rb [hunk 21]
```diff
@@ -1,22 +1,26 @@
...
       end
 
       def ==(o)
         o.kind_of?(GreaterThanOrEqual) &&
           o.obj == @obj &&
-          o.attr == @attr
+          o.key == @key
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/greater_than_or_equal.rb]: The same as in the other cases.

File: lib/ruy/conditions/in.rb [hunk 22]
```diff
@@ -1,22 +1,25 @@
...
-      def initialize(ary, *attrs)
+      # @param ary
+      # @param key
+      # @example check that :key is included in [1,3,5]
+      #   In.new([1,3,5], :key)
+      def initialize(ary, *keys)
         super
         @ary = ary
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/in.rb]: The same as in the other cases.

File: lib/ruy/conditions/in.rb [hunk 23]
```diff
@@ -1,22 +1,25 @@
...
       end
 
       def ==(o)
         o.kind_of?(In) &&
           o.ary == @ary &&
-          o.attr == @attr
+          o.key == @key
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/in.rb]: The same as in the other cases.

File: lib/ruy/conditions/in_cyclic_order.rb [hunk 24]
```diff
@@ -1,20 +1,28 @@
...
+    #
     class InCyclicOrder < Condition
-      attr_reader :from, :to, :attr
+      attr_reader :from, :to, :key
 
-      def initialize(from, to, *attrs)
+      # @param from left bound
+      # @param to right bound
+      # @param key
+      # @example check that :key is between a cycle defined between 100 and -100
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/in_cyclic_order.rb]: THe same as for between a range. I believe that it should be: between 100 and -100 or included in a cycle between 100 and -100.

File: lib/ruy/conditions/in_cyclic_order.rb [hunk 25]
```diff
@@ -1,20 +1,28 @@
...
+      #   InCyclicOrder.new(100, -100, :key)
+      def initialize(from, to, *keys)
         super
         @from = from
         @to = to
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/in_cyclic_order.rb]: The same as in the other cases.

File: lib/ruy/conditions/in_cyclic_order.rb [hunk 26]
```diff
@@ -1,20 +1,28 @@
...
       end
 
       def ==(o)
         o.kind_of?(self.class) &&
           o.from == @from &&
           o.to   == @to &&
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/in_cyclic_order.rb]: The same as in the other cases.

File: lib/ruy/conditions/include.rb [hunk 27]
```diff
@@ -1,28 +1,32 @@
...
-      def initialize(obj, *attrs)
+      # @param obj
+      # @param key
+      # @example check that :key includes 5
+      #   Include.new(5, :key)
+      def initialize(obj, *keys)
         super
         @obj = obj
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/include.rb]: The same as in the other cases.

File: lib/ruy/conditions/include.rb [hunk 28]
```diff
@@ -1,28 +1,32 @@
...
       end
 
       def ==(o)
         o.kind_of?(Include) &&
           o.obj == @obj &&
-          o.attr == @attr
+          o.key == @key
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/include.rb]: The same as in the other cases.

File: lib/ruy/conditions/less_than_or_equal.rb [hunk 29]
```diff
@@ -1,22 +1,26 @@
...
         @obj = obj
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
       end
 
       def ==(o)
         o.kind_of?(LessThanOrEqual) &&
           o.obj == @obj &&
-          o.attr == @attr
+          o.key == @key
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/less_than_or_equal.rb]: The same as in the other cases.

File: lib/ruy/conditions/less_than_or_equal.rb [hunk 30]
```diff
@@ -1,22 +1,26 @@
...
-      def initialize(obj, *attrs)
+      # @param obj
+      # @param key
+      # @example check that :key is less than or equal to 5
+      #   LessThanOrEqual.new(5, :key)
+      def initialize(obj, *keys)
         super
         @obj = obj
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/less_than_or_equal.rb]: The same as in the other cases.

File: lib/ruy/conditions/some.rb [hunk 31]
```diff
@@ -1,27 +1,31 @@
...
 
-      # @param attr Context attribute's name
-      def initialize(*attrs)
+      # @param key
+      # @example check that at least a value from :key matches the sub-conditions
+      #   Some.new(:key)
+      def initialize(*keys)
         super
-        @attr = attrs.first if attrs.any?
+        @key = keys.first if keys.any?
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/some.rb]: The same as in the other cases.

File: lib/ruy/conditions/some.rb [hunk 32]
```diff
@@ -1,27 +1,31 @@
...
       end
 
-      def call(ctx)
-        solve(ctx).any? do |newctx|
-          ctx = Ruy::Context.new(newctx)
+      def evaluate(enum)
+        enum.any? do |ctx|
+          newctx = Ruy::Context.new(ctx)
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/some.rb]: I would prefer new_ctx

File: lib/ruy/conditions/some.rb [hunk 33]
```diff
@@ -1,27 +1,31 @@
...
 
-          Ruy::Utils::Rules.evaluate_conditions(conditions, ctx)
+          Ruy::Utils::Rules.evaluate_conditions(conditions, newctx)
         end
       end
 
       def ==(o)
         super &&
-          o.attr == @attr
+          o.key == @key
```
Comment (ignaciovillaverde) [inline lib/ruy/conditions/some.rb]: The same as in the other cases.

File: lib/ruy/outcome.rb [hunk 34]
```diff
@@ -1,27 +1,29 @@
...
     # It also returns the value when there's no conditions.
     #
     # @param [Ruy::Context] ctx
     #
-    # @return [Object]
-    # @return [nil] If some of the conditions does not succeeds
+    # @return the value
+    # @return [nil] if conditions are not met
     def call(ctx)
       if @root_condition.call(ctx)
```
Comment (ignaciovillaverde) [inline lib/ruy/outcome.rb]: Could be in one line.

File: lib/ruy/outcome.rb [hunk 35]
```diff
@@ -1,27 +1,29 @@
...
       @value = value
       @root_condition = Ruy::Conditions::All.new
     end
 
-    # Returns the value of this outcome is all of the conditions succeed.
-    #
+    # Returns the value of this outcome if all of the conditions succeed.
     # It also returns the value when there's no conditions.
     #
     # @param [Ruy::Context] ctx
```
Comment (ignaciovillaverde) [inline lib/ruy/outcome.rb]: The same as in the other cases.

File: lib/ruy/rule.rb [hunk 36]
```diff
@@ -14,61 +14,67 @@ def initialize
...
     def set(name, value)
       @attrs[name] = value
     end
 
-    # Evaluates rule conditions return the corresponding outcome or fallback depending on whether
-    # the conditions matched or not.
+    # Evaluates that rule conditions return the corresponding outcome
+    # or fallback depending on whether the conditions matched or not
     #
     # @param [Hash] context
```
Comment (ignaciovillaverde) [inline lib/ruy/rule.rb]: The same as in the other cases.

File: lib/ruy/rule.rb [hunk 37]
```diff
@@ -14,61 +14,67 @@ def initialize
...
-    end
-
-    # Defines a memoized value.
+    # Stores the name of a lazy variable whose block will be defined
+    # in the evaluated context
     #
-    # The value will be resolved upon the context during evaluation. Let is lazy evaluated, it is
-    # not evaluated until the first condition referencing it is invoked. Once evaluated, its value
-    # is stored, so subsequent invocations during the same evaluation will resolve again its value.
+    # @param [Symbol] name
```
Comment (ignaciovillaverde) [inline lib/ruy/rule.rb]: The same as in the other cases.

File: lib/ruy/conditions/assert.rb (hunk 2 above)
Comment (migl) [inline lib/ruy/conditions/assert.rb]: This was done on purpose to allow `nil` as a value for a key. `has_key?` checks whether `@key` is defined or not. Setting it to `nil` will make `defined?` return `true` which is not the expected outcome.

File: lib/ruy/conditions/condition.rb (hunk 10 above)
Comment (migl) [inline lib/ruy/conditions/condition.rb]: There's a protected `key` reader defined in `Condition`.

File: lib/ruy/conditions/day_of_week.rb (hunk 11 above)
Comment (migl) [inline lib/ruy/conditions/day_of_week.rb]: We want it to be consistent with all the other files where there are too many keys to make it a one liner.

File: lib/ruy/conditions/day_of_week.rb (hunk 12 above)
Comment (migl) [inline lib/ruy/conditions/day_of_week.rb]: `evaluate` is the most important method of a `Condition` class, and it basically takes charge of performing the evaluation the class is meant to do, so the description of what it does lies in the description of the class. What I can do is to add a `@see` directive to reference `super`, but basically the documentation would be the documentation of the class again.

**Comment by snmgian** [comment]
:+1:

**Pull request merged**